crates/environ/src/module.rs

```diff
@@ -311,6 +272,149 @@ impl Default for MemoryInitialization {
+    ) -> InitMemory {
+        let initializers = match self {
+            // Fall through below to the segmented memory one-by-one
+            // initialization.
+            MemoryInitialization::Segmented(list) => list,
+
+            // If previously switched to paged initialization then pass through
+            // all those parameters here to the `write` callback.
+            //
+            // Note that existence of `Paged` already guarantees that all
+            // indices are in-bounds.
+            MemoryInitialization::Paged { map } => {
+                for (index, pages) in map {
+                    for (page_index, page) in pages {
+                        debug_assert_eq!(page.end - page.start, WASM_PAGE_SIZE);
+                        let result = write(index, *page_index * u64::from(WASM_PAGE_SIZE), page);
+                        if !result.is_ok() {
+                            return result;
+                        }
+                    }
+                }
+                return InitMemory::Ok;
+            }
+        };
+
+        for initializer in initializers {
+            let MemoryInitializer {
+                memory_index,
+                base,
+                offset,
+                ref data,
+            } = *initializer;
+
+            // First up determine the start/end range and verify that they're
+            // in-bounds for the initial size of the memory at `memory_index`.
+            // Note that this can bail if we don't have access to globals yet
+            // (e.g. this is a task happening before instantiation at
+            // compile-time).
+            let base = match base {
+                Some(index) => match get_global(index) {
+                    Some(base) => base,
+                    None => return InitMemory::NoGlobal,
+                },
+                None => 0,
+            };
+            let start = match base.checked_add(offset) {
+                Some(start) => start,
+                None => return InitMemory::OutOfBounds,
+            };
+            let len = u64::try_from(data.len()).unwrap();
+            let end = match start.checked_add(len) {
+                Some(end) => end,
+                None => return InitMemory::OutOfBounds,
+            };
+
+            // Note that this `minimum` can overflow if `minimum` is
+            // `1 << 48`, the maximum number of minimum pages for 64-bit
+            // memories. If this overflow happens, though, then there's no need
+            // to check the `end` value since `end` fits in a `u64` and it is
+            // naturally less than the overflowed value.
+            //
+            // This is a bit esoteric though because it's impossible to actually
+            // create a memory of `u64::MAX + 1` bytes, so this is largely just
+            // here to avoid having the multiplication here overflow in debug
+            // mode.
+            if let Some(max) =
```

Comment:
Could we avoid the conditional-on-overflow logic by testing for `(end / WASM_PAGE_SIZE) > get_cur_page_size(memory_index)` instead, i.e., doing the comparison in the Wasm-page-count domain?

(Or I guess, the above but rounding up, so `((end + WASM_PAGE_SIZE - 1) / WASM_PAGE_SIZE)`, maybe with that wrapped in a helper...)

Reply:
Ah yeah my hope here was to avoid having to deal with rounding logic and being off-by-one by accident. I think this would work if `end` were rounded up but that's still tricky checked arithmetic to do as well.